### PR TITLE
Generate typed PocketBase type

### LINKED ISSUE #81: Typed PocketBase client
In the [latest release](https://github.com/pocketbase/js-sdk/releases/tag/v0.18.3) of the PocketBase JavaScript SDK there was added support for providing a typed instance of the client using type assertion. I think a generic version of this shape could probably be made from the `CollectionResponses` type from this library, but it would probably be a complicated TypeScript mess. It would instead be really cool if this library generated the necessary types automatically.

~I'm raising a PR for this in a few minutes.~
PR opened as #82.

### PR DESCRIPTION
**Pull request by scriptcoded**
In the [latest release](https://github.com/pocketbase/js-sdk/releases/tag/v0.18.3) of the PocketBase JavaScript SDK there was added support for providing a typed instance of the client using type assertion. This PR adds a new exported type that fits this shape and can be used directly with PocketBase like this:

```ts
import { TypedPocketBase } from "./pocketbase-types";

const pb = new PocketBase('...') as TypedPocketBase
```

To make this work I had to add an import statement to the generated file to get the base types from PocketBase. The resulting file looks something like this:

```ts
/**
* This file was @generated using pocketbase-typegen
*/

import PocketBase, { RecordService } from 'pocketbase'

// ...

// Type for usage with type asserted Pocketbase instance
// https://github.com/pocketbase/js-sdk#specify-typescript-definitions

export type TypedPocketBase = PocketBase & {
	collection(idOrName: string): RecordService
	collection(idOrName: 'posts'): RecordService<PostsResponse>
	collection(idOrName: 'users'): RecordService<UsersResponse>
}
```

An alternative if you don't like having this import would be to use generics and have the user pass down the `PocketBase` and `RecordService` types, but I think the user experience would be a lot worse.

Closes #81

### CODE REVIEW

**Comment by patmood** [comment]
I love this, thanks for the contribution. Really glad pocketbase added this and I think it'll make this library much easier to use going forward.

I think there should be a `--no-sdk` command line option for users that just want pure types (maybe they are using the REST api).

Before merging I want to test it out and check some edge cases

**Comment by patmood** [comment]
Im not sure how expanded relation types and the generic fields for JSON objects will work with this setup. That's not a blocker though

**Comment by scriptcoded** [comment]
> I think there should be a --no-sdk command line option for users that just want pure types (maybe they are using the REST api).

I agree this would be a good addition. I'll make sure to add it. I also realize I completely forget to update the documentation so I'll make sure that gets added as well.

As for expanded relations I totally didn't think of that so I'll have to read up on exactly how the types behave in that case before I give my thoughts :smile: In the worst case you should still be able to pass generic types to the collection methods (like `.getOne`) even though you're using the typed SDK.

File: src/collections.ts [hunk 1]
```diff
@@ -31,3 +31,15 @@ export function createCollectionResponses(
...
 }
+
+export function createTypedPocketbase(
+  collectionNames: Array<string>
+): string {
+  const nameRecordMap = collectionNames
+    .map((name) => `\tcollection(idOrName: '${name}'): RecordService<${toPascalCase(name)}Response>`)
+    .join("\n")
+  return `export type TypedPocketBase = PocketBase & {
+\tcollection(idOrName: string): RecordService
```
Comment (scriptcoded) [inline src/collections.ts:42]: Do you think this "fallback" should be kept? By having it here the user can pass in collection names that are not generated by `pocketbase-typegen`. That can be seen as both a good and a bad thing, as by having it there is no type checking for invalid collection names.

Comment (patmood) [reply src/collections.ts:42]: Good point. I would remove it since we have the complete list of collections. An type error is a good hint that they need to regenerate

File: src/lib.ts [hunk 2]
```diff
@@ -49,9 +57,13 @@ export function generate(results: Array<CollectionRecord>): string {
...
     createCollectionRecords(sortedCollectionNames),
     createCollectionResponses(sortedCollectionNames),
+    options.sdk && TYPED_POCKETBASE_COMMENT,
+    options.sdk && createTypedPocketbase(sortedCollectionNames)
   ]
 
-  return fileParts.join("\n\n")
+  return fileParts
+    .filter(Boolean)
+    .join("\n\n") + '\n'
```
[the remark this answers is not in the document]
Comment (patmood) [reply src/lib.ts:66]: Thanks!

File: README.md [hunk 3]
```diff
@@ -71,15 +73,27 @@ The output is a typescript file `pocketbase-types.ts` ([example](./test/pocketba
   - `[CollectionName][FieldName]Options` If the collection contains a select field with set values, an enum of the options will be generated.
 - `CollectionRecords` A type mapping each collection name to the record type.
 - `CollectionResponses` A type mapping each collection name to the response type.
+- `TypedPocketBase` A type for usage with type asserted PocketBase instance.
 
 ## Example Usage
 
+In [PocketBase SDK](https://github.com/pocketbase/js-sdk) v0.18.3+ you can use the PocketBase SDK with type assertion to have collections automatically typed:
```
Comment (patmood) [inline README.md:80]:
```suggestion
Using PocketBase SDK v0.18.3+, collections can be [automatically typed](https://github.com/pocketbase/js-sdk#specify-typescript-definitions) using the generated `TypedPocketBase` type:
```
Slight wording suggestion

File: README.md [hunk 4]
```diff
@@ -71,15 +73,27 @@ The output is a typescript file `pocketbase-types.ts` ([example](./test/pocketba
...
+```typescript
+import { TypedPocketBase } from "./pocketbase-types"
+
+const pb = new PocketBase('http://127.0.0.1:8090') as TypedPocketBase
+
+await pb.collection('tasks').getOne("RECORD_ID") // -> results in TaskResponse
+await pb.collection('posts').getOne("RECORD_ID") // -> results in PostResponse
+```
+
 In [PocketBase SDK](https://github.com/pocketbase/js-sdk) v0.8+ you can use generic types when fetching records, eg:
```
Comment (patmood) [inline README.md:91]:
```suggestion
Alternatively, you can use generic types for each request, eg:
```
Nobody new will be using <v0.8 now

**Review by patmood — COMMENTED** [review]
Thanks for doing this! I'm traveling at the moment so I appreciate the update for the new SDK. Could you also bump the version in `package.json` to 1.1.14?

File: README.md (hunk 3 above)
Comment (scriptcoded) [reply README.md:80]: Much better :smile:

File: src/constants.ts [hunk 5]
```diff
@@ -1,9 +1,11 @@
 export const EXPORT_COMMENT = `/**
 * This file was @generated using pocketbase-typegen
 */`
+export const IMPORTS = `import PocketBase, { RecordService } from 'pocketbase'`
```
[the remark this answers is not in the document]
Comment (scriptcoded) [reply src/constants.ts:4]: Had to import in two lines due to this:
> A type-only import can specify a default import or named bindings, but not both.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html

**Comment by scriptcoded** [comment]
@patmood I don't know what versioning strategy you're using, but seeing as this adds additional functionality, should this maybe me a minor bump i.e. 1.2.0?

**Comment by patmood** [comment]
Thanks again @scriptcoded - Looks good! 🎉 I'll update to 1.2.0 before I release

**Comment by anbraten** [comment]
Just noticed this one when updating to the latest release. It's such a useful nifty feature Thanks ❤️
